pepr/attack_runner.py

```diff
@@ -0,0 +1,208 @@
+                # Parse MIA configuration
+                model_function = functions[yaml_attack_pars["create_compile_model"]]
+
+                yaml_data_conf = data["data_conf"][i]
+                reference_indices = list(
+                    np.load(yaml_data_conf["reference_indices_path"])
+                )
+                target_indices = list(np.load(yaml_data_conf["target_indices_path"]))
+                evaluation_indices = list(
+                    np.load(yaml_data_conf["evaluation_indices_path"])
+                )
+                record_indices_per_target = np.load(
+                    yaml_data_conf["record_indices_per_target_path"]
+                )
+
+                target_models = []
+                for path in yaml_attack_pars["target_model_paths"]:
+                    target_models.append(tf.keras.models.load_model(path))
+
+                # Setup dictionaries for attack
+                attack_pars = {
+                    "number_classes": yaml_attack_pars["number_classes"],
+                    "number_reference_models": yaml_attack_pars[
+                        "number_reference_models"
+                    ],
+                    "reference_training_set_size": yaml_attack_pars[
+                        "reference_training_set_size"
+                    ],
+                    "create_compile_model": model_function,
+                    "reference_epochs": yaml_attack_pars["reference_epochs"],
+                    "reference_batch_size": yaml_attack_pars["reference_batch_size"],
+                    "hlf_metric": yaml_attack_pars["hlf_metric"],
+                    "hlf_layer_number": yaml_attack_pars["hlf_layer_number"],
+                    "number_target_records": yaml_attack_pars["number_target_records"],
+                }
+
+                data_conf = {
+                    "reference_indices": reference_indices,
+                    "target_indices": target_indices,
+                    "evaluation_indices": evaluation_indices,
+                    "record_indices_per_target": record_indices_per_target,
+                }
+
+                # Create attack object
+                logger.info(f"Attack: {yaml_attack_pars['attack_alias']}")
+                attack = gmia.DirectGmia(
+                    yaml_attack_pars["attack_alias"],
+                    attack_pars,
+                    np.load(yaml_attack_pars["path_to_dataset_data"]),
+                    np.load(yaml_attack_pars["path_to_dataset_labels"]),
+                    data_conf,
+                    target_models,
+                )
+
+                attack.run()
+                attack.create_attack_section(report_save_path)
+                sections.append(attack.report_section)
+
+                logger.debug("Serialize attack object.")
+                pickle_attack_obj(attack)
+
+        logger.info("Generate final report.")
+        report.report_generator(report_save_path, sections, pdf=pdf)
```

Comment:
To me it makes sense to separate the attack runner from the report generator. I would expect the attack runner to be executed almost as a drop-in replacement for the simple attack.run() function. What do you think? :)

Reply:
Yep, I fully agree. The attack runner is now split in two functions: 
- `run_attacks` which parses the config file and runs the attacks
- `create_report` which creates one report out of a list of attack objects

Reply:
Nice 💯